sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_ng_udp
+ * @{
+ *
+ * @file
+ * @brief       UDP implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @}
+ */
+
+#include <stdint.h>
+#include <errno.h>
+
+#include "kernel.h"
+#include "byteorder.h"
+#include "msg.h"
+#include "thread.h"
+#include "net/ng_udp.h"
+#include "net/ng_netbase.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   Save the UDP's thread PID for later reference
+ */
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+
+/**
+ * @brief   Allocate memory for the UDP thread's stack
+ */
+static char _stack[NG_UDP_STACK_SIZE];
+
+/**
+ * @brief   Calculate the UDP checksum dependent on the network protocol
+ *
+ * @note    If the checksum turns out to be 0x0000, the function returns 0xffff
+ *          as specified in RFC768
+ *
+ * @param[in] pkt           pointer to the packet in the packet buffer
+ * @param[in] pseudo_hdr    pointer to the network layer header
+ * @param[in] payload       pointer to the payload
+ *
+ * @return                  the checksum of the pkt
+ * @return                  0 on error
+ */
+uint16_t _calc_csum(ng_pktsnip_t *hdr, ng_pktsnip_t *pseudo_hdr,
+                    ng_pktsnip_t *payload)
+{
+    switch (pseudo_hdr->type) {
+#ifdef MODULE_NG_IPV6
+        case NG_NETTYPE_IPV6:
+            return _calc_csum_ipv6(hdr, pseudo_hdr, payload);
+#endif
+        default:
+            return -ENOENT;
+    }
+}
+
+uint16_t _calc_csum_ipv6(ng_pktsnip_t *udp, ng_pktsnip_t *ip,
+                         ng_pktsnip_t *payload)
```

Comment:
ups, I was completely off - sometimes you wonder what you were thinking at a certain moment... thx for pointing this out!